pkg/virt-launcher/virtwrap/sriov/pcipool_test.go

```diff
@@ -0,0 +1,136 @@
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package sriov_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/sriov"
+)
+
+var _ = Describe("SRIOV PCI address pool", func() {
+	It("Fails to pop an address given no interfaces", func() {
+		pool := sriov.NewPCIAddressPool([]v1.Interface{})
+		expectPoolPopFailure(pool, "foo")
+	})
+
+	It("fails to pop an address given a non-SRIOV interface", func() {
+		pool := sriov.NewPCIAddressPool([]v1.Interface{{Name: "foo"}})
+		expectPoolPopFailure(pool, "foo")
+	})
+
+	It("fails to pop an address given a missing resource name env", func() {
+		iface := newSRIOVInterface("foo")
+		pool := sriov.NewPCIAddressPool([]v1.Interface{iface})
+
+		expectPoolPopFailure(pool, "foo")
+	})
+
+	It("fails to pop an address given a missing device name env", func() {
+		net := newNetworkData("net1", "resource1", "0000:81:01.0")
+		env := []envData{net.ResourceEnv}
+		withEnvironmentContext(env, func() {
+			iface := newSRIOVInterface(net.Name)
+			pool := sriov.NewPCIAddressPool([]v1.Interface{iface})
+
+			expectPoolPopFailure(pool, net.Name)
+		})
+	})
+
+	It("provides 1 address given 1xInterface, 1xResource, 1xPCI", func() {
+		net := newNetworkData("net1", "resource1", "0000:81:01.0")
+		env := []envData{net.ResourceEnv, net.DeviceEnv}
+		withEnvironmentContext(env, func() {
+			iface := newSRIOVInterface(net.Name)
+			pool := sriov.NewPCIAddressPool([]v1.Interface{iface})
+
+			Expect(pool.Pop(net.Name)).To(Equal(net.PCIAddresses[0]))
+			expectPoolPopFailure(pool, net.Name)
+		})
+	})
+
+	It("provides 2 addresses given 1xInterface, 1xResource, 2xPCI", func() {
+		net := newNetworkData("net1", "resource1", "0000:81:01.0", "0000:81:01.1")
+		env := []envData{net.ResourceEnv, net.DeviceEnv}
+		withEnvironmentContext(env, func() {
+			iface := newSRIOVInterface(net.Name)
+			pool := sriov.NewPCIAddressPool([]v1.Interface{iface})
+
+			Expect(pool.Pop(net.Name)).To(Equal(net.PCIAddresses[0]))
+			Expect(pool.Pop(net.Name)).To(Equal(net.PCIAddresses[1]))
+			expectPoolPopFailure(pool, net.Name)
+		})
+	})
+
+	It("provides 2 addresses given 2xInterface, 1xResource, 2xPCI", func() {
+		net1 := newNetworkData("net1", "resource1", "0000:81:01.0", "0000:81:02.0")
+		net2 := newNetworkData("net2", "resource1")
```

Comment:
This is pretty weird, we need a new struct just for the resources and share it between net1 and net2.

```golang
resource1 := newResourceData("resource1", "0000:81:01.0", "0000:81:02.0")
net1 := newNetworkData("net1", resource1)
net2 := newNetworkData("net2", resource1)
```

Reply:
I think it is, from first glance at the test you don't see that there is a relation between net1 and net2 you realize about it after you dig into the implementation test has to be as declarative as possible and expressing them with proper data types helps with that.